### ISSUE TITLE
Add variable type conversion in BUFR converter so the result variable type follows the IODA convention

### ISSUE THREAD

**Issue by emilyhcliu** [JCSDA-internal/ioda-converters#965]
### In current BUFR converter, the output IODA variables inherit the data types queried from the BUFR.  These data types from BUFR are mostly float and string. There is a need to add the variable type conversion capability in BUFR converter so that the output data types follow the IODA convention.

The IODA data type can be found in the [Data Storage Types](https://docs.google.com/spreadsheets/d/1fHb2QIHOwgp4d6plyn9Dj9HZFZsDRxcPaUV6hdTb5BU/edit#gid=741804046) section in the [IODA Convention Table](https://docs.google.com/spreadsheets/d/1fHb2QIHOwgp4d6plyn9Dj9HZFZsDRxcPaUV6hdTb5BU/edit?usp=sharing).

For example: 

This is the output IODA variable (sensorChannelNumber) from BUFR.  The data type queried from BUFR is **float**. The IODA convention requires **uint32_t** ( a specific integer type)
````
        float sensorChannelNumber(Location, Channel) ;
                sensorChannelNumber:_FillValue = 1.e+11f ;
                string sensorChannelNumber:long_name = "Sensor Channel Number" ;
```

**Can we add a "type" keyword in the IODA part of the YAML to specify the data type to be converted to?**

```
        - name: "MetaData/sensorChannelNumber"
          source: variables/sensorChannelNumber
          longName: "Sensor Channel Number"
          type: uint32
```

Or 

the BUFR export part:
```
          sensorChannelNumber:
            query: "*/IASICHN/CHNM"
            type: uint32
```


## Requirements

## Acceptance Criteria (Definition of Done)

## Dependencies
This issue has a higher priority than [issue #394](https://github.com/JCSDA-internal/ioda-converters/issues/964)

**Comment by emilyhcliu** [comment]
> So I think the right way to fix this would be to read the BUFR TABLE B data and grab the UNITS for the mnemonic. This way the user wouldn't need to indicate which fields to turn into int or float types (there are a lot of them). Unfortunately I would need to extend NCEPLIB-bufr a little to accomplish this... I can know the type automatically. What do you think?

@rmclaren what you proposed is even better since the types can be automatically determined internally.

**Comment by srherbener** [comment]
My understanding is that we want the variables in the output ioda file to follow the data types specified in the Conventions, which is not necessarily the data type inferred from the BUFR units. While it might be helpful to infer the data type from the BUFR units, I think we still need to have a mechanism in place to ensure that the output variable data types are following the JEDI conventions.

I believe that @rhoneyager is working on a scheme to import the JEDI conventions so that they can be used to enforce the convention names, data types, etc. If I've got that right, we should coordinate with @rhoneyager's work to address this issue.

**Comment by emilyhcliu** [comment]
@rmclaern  Yes.  Let's make plans and prioritize work that needs to be done based on your capacity and schedule.   We could put the data type work in the to-do list, This to-do list should be included in the documentation so the users know what to expect in the next release. We already have a few on the to-do list.
The following issues document the desired features:
- [ ] [Issue #964](https://github.com/JCSDA-internal/ioda-converters/issues/964)   ----> enhance the "transform" functionality to handle variable transform beyond constant offset/scaling and variable transform involving other variables
- [ ] [Issue #965](https://github.com/JCSDA-internal/ioda-converters/issues/965)   ----> ensure variable type from bufr follows the JEDI convention
- [ ] [Issue #967 ](https://github.com/JCSDA-internal/ioda-converters/issues/967)  ----> flatten dimension [Location, 1] to dimension[Location]

**Comment by rmclaren** [comment]
I added a new epic for the next enhancements we want to implement (after we merge our current work into develop and close the currently open EPIC). Please move "future" things to this new epic. Things we plan to defer in the current epic should be moved to the new one as well (?).

https://app.zenhub.com/workspaces/jedi-5eb436d0a544107acce79780/issues/jcsda-internal/ioda-converters/969

@emilyhcliu  @srherbener Is this the right approach to take?

**Comment by rmclaren** [comment]
@srherbener We could always provide the means to override whatever the BUFR unit indicates. I'm guessing the majority of time the implied type will be ok (?). What do you think?

**Comment by srherbener** [comment]
> I added a new epic for the next enhancements we want to implement (after we merge our current work into develop and close the currently open EPIC). Please move "future" things to this new epic. Things we plan to defer in the current epic should be moved to the new one as well (?).
> 
> https://app.zenhub.com/workspaces/jedi-5eb436d0a544107acce79780/issues/jcsda-internal/ioda-converters/969
> 
> @emilyhcliu @srherbener Is this the right approach to take? I'm trying to seperate the things we want right now, from the things we want to add as enhancements in the future.

I think that sounds good. If you have a number of tasks you want to work on now for the merge it would probably be good to create another epic for tracking those tasks.

**Comment by rmclaren** [comment]
@srherbener I guess I was thinking this epic had that role:

https://app.zenhub.com/workspaces/jedi-5eb436d0a544107acce79780/issues/jcsda-internal/ioda-converters/425

**Comment by rmclaren** [comment]
@srherbener You are correct about the BUFR Table indicating the number of bits used to encode any particular value. The thing is that NCEPLIB-bufr converts all these values to the same representation (32 bit floats in our case) so the number isn't too helpful (unless we want to concatenate the values). I don't know how far we want to take this...

**Comment by rmclaren** [comment]
Merging this issue is dependent on the following PR:

https://github.com/NOAA-EMC/NCEPLIBS-bufr/pull/222

**Comment by srherbener** [comment]
> Merging this issue is dependent on the following PR:
> 
> [NOAA-EMC/NCEPLIBS-bufr#222](https://github.com/NOAA-EMC/NCEPLIBS-bufr/pull/222)

@rmclaren does this mean that we need this NCEPLIBS-bufr PR merged, a tag assigned to a version containing that merge, and switch our stack over to that tag before we can merge your feature/query_cxx branch to develop? Just want to know what to plan for. Thanks!

**Comment by rmclaren** [comment]
@srherbener Yes, that is exactly what needs to happen:) Thanks Steve!

**Comment by rmclaren** [comment]
@srherbener Depending on how things work out timing wise we might end up merging feature/query_cxx to develop so this one would then just get merged into develop directly.

**Comment by srherbener** [comment]
> @srherbener Depending on how things work out timing wise we might end up merging feature/query_cxx to develop so this one would then just get merged into develop directly.

FYI: If you create a feature branch for this before feature/query_cxx gets merged, you can base the new feature branch on feature/query_cxx (instead of develop). Github will keep track of the base branch and update it accordingly. Say the feature branch for this issue is called feature/var_type_convert and you base it on feature/query_cxx. You can submit a PR for feature/var_type_convert and set that up to merge into feature/query_cxx. Then let's say feature/query_cxx gets merged to develop (before feature/var_type_convert gets merged into feature/query_cxx). When the merge of feature/query_cxx into develop occurs, github will automatically change the base of the feature/var_type_convert PR to develop.

**Comment by rmclaren** [comment]
Yup, that's exactly what I did :) Its more a question of minimal features people want in the first release.

**Comment by rmclaren** [comment]
This functionality has been implemented in feature/query_cxx_types. Will update with PR when its issued (currently waiting on updating all the test files (yaml and output files).
